### ISSUE TITLE
[REQUEST] be able to switch languages when wanting to type in search

### ISSUE THREAD

**Issue by IHaveAidsFromMia** [iBicha/playlet#362]
As the title says, is it also possible to add piped instances (in case, invidious servers get down or get sued) (youtube did try to sue invidious devs).

Furthermore, is it also possible to be able to switch a language when typing in the search? (unless...it is already present as a feature and i haven't noticed?). 

Keep up the work! :)

**Comment by iBicha** [comment]
Thanks for the suggestions
We already have an issue for piped #75 

Which language are you looking for support for? Roku has very limited language support, and doesn't display them correctly except for a select few. For example, these are the special characters that can be displayed beyond the english/latin alphabet:
<img width="631" alt="keyboard" src="https://github.com/iBicha/playlet/assets/17722782/feaa8981-e762-4929-9b32-f4efb742431c">
Other languages (like Asian languages, Arabic, Farsi, and so on) are not supported by Roku.

Your second best option is to use the web app and/or YouTube to search and then cast.

**Comment by IHaveAidsFromMia** [comment]
Yeah...i've noticed in a glimpse that piped had some issues for roku, sadly.

For the language support, i thought it would've been possible to add some cyrillic characters. Though, if it's an issue from Roku themselves. Then yeah...i can see the issue.

Oh well...it doesn't really stop me by using cyrillic by typing it in english/latin characters. 

Sorry, for bothering. I was just curious if it was possible.

**Comment by iBicha** [comment]
> Yeah...i've noticed in a glimpse that piped had some issues for roku, sadly.

Oh I did not mean Playlet supports Piped, I meant we have a Github issue for it

> For the language support, i thought it would've been possible to add some cyrillic characters. Though, if it's an issue from Roku themselves. Then yeah...i can see the issue.

I believe cyrillic can work - the languages covered are listed here, and these can be rendered https://github.com/iBicha/playlet/blob/151d05831cb92cc5f1fe41d99d2b41393248d3c5/playlet-lib/src/config/ISO-639-1-to-ISO-639-2T.json5#L4-L11

So technically the keyboard can have a button that opens a dialog, and then a different keyboard layout can be chosen.

**Comment by ND-451** [comment]
Srry for duplicate and im new to roku apps but im guessing this is sideload only and i have to add another language into the keyboard config?

**Comment by ND-451** [comment]
> No problem, but no, the keyboard located at the search is currently not configurable.
Is this for store version or manual install via dev mode or both?
> Which language are you looking to support?
I was thinking it would be nice to get the basic out of the way but the one specifically im looking for is russian i can use screen cast on youtube but it a hassle thing

**Comment by ND-451** [comment]
A idea i had since (i tested this and it works) you can have 2 instances of playlet installed, one via store and one via dev mode sideload so theoretically you could leave the store in English and modify the zip for the sideload one to contain another language.

Its not perfect but it is a quick fix until a more permanent solution but it would require user to have some technical experience to configure the zip before installing it but then again they are on github so unlikely (i would like to think)

**Comment by iBicha** [comment]
Feature released in 0.42.0
Initial list: English, Deutsch, Español, Français, Português and Русский
This list is arbitrary, and more can be added later. I had to start somewhere.

The process of getting languages right is tedious, especially since each language has it's own special considerations.
I don't know if I should work on more languages, at least not without a feature request, because I'll probably mess it up, since I'm no expert. I wouldn't know if I'm missing characters, or if the layout is not optimal for how you usually write in any particular language. So, anyone is welcome to ask for the language they need, or even open a pull request, and I'll be happy to assist.

That being said, I'll document a bit in this comment where the languages are defined:
- First, the languages are setup in the keyboard component: https://github.com/iBicha/playlet/blob/v0.42.0/playlet-lib/src/components/Screens/SearchScreen/SearchKeyboard/SearchKeyboard.bs#L10
- Then, there's a list of them in the settings file https://github.com/iBicha/playlet/blob/v0.42.0/playlet-lib/src/config/preferences.json5#L205-L224 this is what allows the preference to be saved
- Finally, each language is defined in it's own file https://github.com/iBicha/playlet/tree/v0.42.0/playlet-lib/src/config/search-kb-layout

And that's basically it. Adding a new language is about adding a new file in the `search-kb-layout` folder, and then updating the list of languages. Added languages are subject to text rendering limitations https://github.com/iBicha/playlet/issues/229
